DEP 44 - Clarified release process (#78)
django/deps: draft/0044-clarify-release-process.rst

=================================
DEP 0044: Clarify Release Process
=================================

> :DEP: 0041

Reviewer: DEP 44

:Author: Carlton Gibson
:Implementation Team: Carlton Gibson
:Shepherd: Mariusz Felisiak
:Status: Draft
:Type: Process
:Created: 2022-11-03
:Last-Modified: 2022-11-15

.. contents:: Table of Contents
   :depth: 3
   :local:

Abstract
========

This DEP clarifies Django's time-based release process, providing additional
information to the 2015 DEP 4 that established the release process.

Specification
=============

Django has a time-based release schedule, as specified in `DEP 4`__ and the `Release
Process`__ page of the Django documentation. Supplemental information is available
in the `How is Django Formed?`__ release process checklist, and the `2015 post on
the Django blog announcing the then new release schedule`__.

__ https://github.com/django/deps/blob/main/final/0004-release-schedule.rst
__ https://docs.djangoproject.com/en/dev/internals/release-process/
__ https://docs.djangoproject.com/en/dev/internals/howto-release-django/
__ https://www.djangoproject.com/weblog/2015/jun/25/roadmap/

Scheduling is entirely mechanical. Django has a fixed eight monthly release

> cycle for major releases — Apr - Dec - Aug, and back to Apr over a 24month

Reviewer 2: It could be a correct English form, I've just never seen it worded that way

```suggestion
cycle for major releases — Apr - Dec - Aug, and back to Apr over a 24 month
```

> cycle for major releases — Apr - Dec - Aug, and back to Apr over a 24 month
> period.

Reviewer 3: ```suggestion
cycle for major releases: April, December, August, and back to April over a 24
month period.
```

Final releases are made early in the first week of the month. Major releases
have a pre-release phase, which has one alpha release, one beta, and one
release candidate.

* The release candidate comes two weeks before the final release.
* The beta comes four weeks before the release candidate.
* The alpha comes five weeks before the beta.

The exact day is left to the Releaser's discretion.

.. note::

    Factors influencing exact release day

    Factors that are considered when setting the exact release day include:
    being as close to the 1st of the month as possible, releaser availability,

>     avoiding Fridays and weekends, avoiding days such as April 1st, and the

Reviewer 3: can we be more specific about why we would avoid April 1?

```suggestion
    avoiding Fridays and weekends, avoiding April 1st (April Fool’s Day), and the
```

    largest international public holidays, such as New Year's Day.

    In general, other public holidays cannot be considered as there are too
    many internationally. Historically, though, releasers have tried to avoid
    July 4th for the sake of the US holiday there.

    This note is purely informational.

The Releaser acting as release manager for the next major version determines
the concrete dates for this schedule in the week following the previous major
version release. The reality is this falls to one of the Fellows, who by
convention alternate the release manager role for each major version.

> As a matter of course regular contributors, including technical board members,
> but not as a matter of obligation, review the proposed schedule in order to
> spot any errors.

Reviewer 3: Reword to make it gramatically clearer, plus I suggest “problems” instead of “errors”, since we might spot something like “oh that is not a good day to release because there is a pre-announced Python zero day fix as well”

```suggestion
As a matter of course, but not as a matter of obligation, regular contributors,
including technical board members, review the proposed schedule in order to
spot any problems.
```

it was not part of DEP 10's scope to address this.

DEP 10 concerned the dissolving of the old Django Core, the establishment of
the Technical Board, and voting around that, together with Merger, Releaser and
other related roles. Adjusting the release process was not mentioned in the
motivation or rationale for DEP 10, and it claims no relevant backwards
incompatible changes.

Likely due to focus elsewhere in DEP 10, the implicit adjustment to the release
process was not caught in review but should be corrected here. If a genuine
desire to change the release process exists, a separate DEP making that case
can be presented.

On the specific points of the original DEP 10 phrasing on the release process,
the release schedule is set entirely mechanically, as per the specification
above. There is no determination needed, and so the requirement for the
technical board to do so introduces artificial busy-work.

Backwards Compatibility
=======================

There is no backwards incompatibility. The change merely brings the wording of
DEP 10 into line with the established release process.

Reference Implementation
========================

> This DEP is it's own implementation. The PR adding it will make necessary

Reviewer 3: ```suggestion
This DEP is its own implementation. The PR adding it will make necessary
```
